Improve submodule commands

tarsius opened the pull request:
I am opening a pr for your branch so that I can comment on some things without risking that those comments are lost.

(Ref #2657)

tarsius commented:
Is it no longer possible to comment on a *commit* itself instead of on changed lines?!

> eec3a3dbd439338b2a5b79586a52ee497edd2609 magit-submodule-add: fix interactive form

The things that seem dubious to you are all due to my desire to add an existing repository (which is already located inside an outer repository) as a submodule. I.e. when I create a new Emacs package, then I first create a new local repository, and then create and push an initial version. Only after that I add the repository as a submodule to `~/.emacs.d`. (I don't think you can add a repository that doesn't have an initial commit yet as a submodule.)

> removed let-binding of default-directory around (magit-get "remote.origin.url") due to reordering of evaluation.  But I don't understand what that was good for anyway.

I just configured "origin" inside the inner repository. Now that I add it as a submodule I don't want to type it again. Changing the minibuffer order breaks that (and that might be okay, see below). But if we do change the order, then we probably should drop the `(magit-get "remote.origin.url")` altogether. (It would be nice to get some "templates" like `git@github.com:my/.git` as completion candidates, but that's something that should be solved in a more general way).

> default value for submodule name has [...code...]

If you add a submodule and then unstage it but later want to add it as a submodule again, then it is nice not to have to type the url again. This happened a lot to me during the creation of [`emacs.g`](https://github.com/emacscollective/emacs.g). I.e. I added a package submodule early on, but later decided that this package should not be part of the proto-starter-kit I publish but should be added later in the private part which is built on top of `emacs.g`, but not public.

---

It probably makes sense to change the *command* `magit-submodule-add` as you suggest. At least the order in which values are read - the existing order doesn't make much sense if you don't know about the above use-case. And I simply use <kbd>s</kbd> to do this now.

![innocence_195](https://cloud.githubusercontent.com/assets/25046/15672523/6d3ec56c-2731-11e6-94e3-027030cdc3c2.jpg)

However, `magit-stage` ultimately adds a submodule by calling `magit-submodule-add` interactively, so all code removed from the latter has to be moved somewhere else.

tarsius commented on lisp/magit-submodule.el:
```diff
@@ -131,10 +131,15 @@ For each section insert the path and the output of `git describe --tags'."
-                (when (string-match-p "\\`[0-9]" it)
-                  (insert ?\s))
-                (insert it))
+              (if (not (file-exists-p ".git")) (insert " (unitialized)")
```
s/unitialized/uninitalized/

tarsius commented:
> a434b0f06870e6fdcb3dd64179389e8a8541915f Add new module section type

This has the potential of breaking many commands.

`magit-submodule-visit` is unnecessary because I already taught `magit-diff-visit-file` about the submodule special case, just press <kbd>RET</kbd>. Well, `magit-submodule-visit` does a little more, but `magit-diff-visit-file` could be changed to call this function without having to add a new section type and then having to make adjustments in many places.

(In the long run I want to use Eieio, among other things so that a section type can derive from another. But in the short run adding a new section type is risky and likely to cause more work than expected.)

npostavs commented:
> This has the potential of breaking many commands.

Speaking only of module sections under the "Modules" heading, not staged/unstaged changes to submodules (I've not looked into changing the latter yet):

Currently `a` (`magit-apply`) and `v` (`magit-reverse`) throw errors on submodule sections.
`s` (`magit-stage`), `u` (`magit-unstage`), and `k` (`magit-discard`) silently do nothing.

As far I can tell, `RET` (`magit-diff-visit-file`) is the only non-broken command anyway.

tarsius commented:
> Speaking only of module sections under the "Modules" heading, not staged/unstaged changes to submodules (I've not looked into changing the latter yet):

I didn't realize that, and retract my objection.

> Currently `a` (`magit-apply`) and `v` (`magit-reverse`) throw errors on submodule sections.
> `s` (`magit-stage`), `u` (`magit-unstage`), and `k` (`magit-discard`) silently do nothing.

And your changes result in more meaningful errors? That would be an improvement.

npostavs commented:
> And your changes result in more meaningful errors? That would be an improvement.

With my changes `a`, `v`, `s`, `u` all prompt for a file/commit, i.e. the submodule section no longer affects these commands. `k` prompts for submodule to `deinit` (defaulting to submodule at point).

---

> The things that seem dubious to you are all due to my desire to add an existing repository (which is already located inside an outer repository) as a submodule[...]

Oh, I hadn't even considered that possibility at all.

> However, `magit-stage` ultimately adds a submodule by calling `magit-submodule-add` interactively, so all code removed from the latter has to be moved somewhere else.

I think it should go into a separate command `magit-submodule-add-from-nested-repo`. And `magit-submodule-add` should be changed to run async like the other `magit-submodule-*` commands (`magit-submodule-add-from-nested-repo` would stay synchronous because `magit-stage` is synchronous).

tarsius commented:
>> However, `magit-stage` ultimately adds a submodule by calling `magit-submodule-add` interactively, so all code removed from the latter has to be moved somewhere else.

> I think it should go into a separate command `magit-submodule-add-from-nested-repo`. And `magit-submodule-add` should be changed to run async like the other `magit-submodule-*` commands (`magit-submodule-add-from-nested-repo` would stay synchronous because `magit-stage` is synchronous).

I think this should be kept as it is, i.e. <kbd>s</kbd> on a nested repository prompts to add that as a submodule instead of calling `git add <repo>`, because that would add the files inside `<repo>` to the outer repository and they would then be tracked in two repositories which wouldn't make sense.

npostavs commented:
> I think this should be kept as it is, i.e. <kbd>s</kbd> on a nested repository prompts to add that as a submodule instead of calling `git add <repo>`, because that would add the files inside `<repo>` to the outer repository and they would then be tracked in two repositories which wouldn't make sense.

Yes agreed, my comment was only regarding where to put the code, not to change the behaviour of `s`.

tarsius commented:
If cherry-picked "magit-submodule-setup: only setup uninitialized" with some cosmetics, and pushed the rest after rebasing onto `master`.

tarsius commented:
We still have to take care of https://github.com/magit/magit/issues/2657#issuecomment-237976396

> `k` on a *staged* submodule attempts `delete-file` on the directory resulting in a `file-error` ("is a directory").

npostavs commented:
> `k` on a *staged* submodule attempts `delete-file` on the directory resulting in a `file-error` ("is a directory").

I think fixing this would require making submodule changes to be a new type of section though.

tarsius commented:
I've applied the branch column patch and implemented the suggested dynamic column width in a separate commit.

Then I have rebased this branch and changed the name of the new section type to `submodule` and moved the respective commit before the others.

I have some doubts about the configuration subpopup and the edit file commands. I haven't yet looked at the other new command.
